src/nitreadme.nit

```diff
@@ -0,0 +1,119 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Helpfull features about README files
+module nitreadme
+
+import frontend
+
+redef class ToolContext
+	# --expand
+	var opt_expand = new OptionBool("Expand packages in their own directory", "--expand")
+
+	# README handling phase
+	var readme_phase: Phase = new ReadmePhase(self, null)
+
+	redef init do
+		super
+		option_context.add_option(opt_expand)
+	end
+end
+
+private class ReadmePhase
+	super Phase
+
+	redef fun process_mainmodule(mainmodule, mmodules) do
+		var expand_packages = toolcontext.opt_expand.value
+		var model = toolcontext.modelbuilder.model
+
+		for mpackage in model.mpackages do
+
+			# Fictive and buggy packages are ignored
+			if not mpackage.has_source then
+				toolcontext.warning(mpackage.location, "no-source",
+					"Warning: `{mpackage}` has no source file")
+				continue
+			end
+
+			# Expand packages
+			if not mpackage.is_expanded then
+				if expand_packages then
+					var path = mpackage.expand
+					toolcontext.info("{mpackage} moved to {path}", 0)
+				else
+					toolcontext.warning(mpackage.location, "no-dir",
+						"Warning: `{mpackage}` has no package directory")
+					continue
+				end
+			end
+		end
+	end
+end
+
+redef class MPackage
+
+	# Has `self` a concrete file?
```

Comment:
Phrasing is strange, "Does `self` have a source file" would be more appropriate I think